### PR TITLE
feat: extract brain area anatomy from NWB location fields

### LINKED ISSUE #1806: extract asset brain area from location field
NWB files have location fields:

- `ImagingPlane.location` (optical physiology)
- `electrodes.location` (extracellular electrophysiology)
- `IntracellularElectrode.location` (intracellular electrophysiology)

These "locations" are brain areas, and we can populate asset-level metadata about brain area by scanning them. They often correspond exactly to Allen CCF brain area abbreviations, and can therefore be mapped directly to the MBAO ontology. See https://github.com/bendichter/analyze-locations/tree/main for details.

### PR DESCRIPTION
**Pull request by bendichter**
## Summary

Closes #1806

- Extracts brain location values from NWB files (`ImagingPlane.location`, `electrodes.location`, `IntracellularElectrode.location`)
- Matches locations against bundled Allen Mouse Brain Atlas CCF structures (~94KB JSON, 1,327 structures)
- Populates `BioSample.anatomy` within the `wasDerivedFrom` field using MBAO identifiers
- Only applies to mouse species (`NCBITaxon_10090`); non-mouse species are skipped

### New files
- `dandi/data/allen_ccf_structures.json` — bundled Allen CCF structures
- `dandi/data/generate_allen_structures.py` — regeneration script (`python -m dandi.data.generate_allen_structures`)
- `dandi/metadata/brain_areas.py` — parsing and matching module
- `dandi/tests/test_brain_areas.py` — 29 unit tests

### Modified files
- `dandi/pynwb_utils.py` — `_get_brain_locations()` extracts location strings from NWB
- `dandi/metadata/util.py` — `extract_wasDerivedFrom()` now adds anatomy to deepest BioSample
- `dandi/tests/test_metadata.py` — 3 integration tests for anatomy in wasDerivedFrom

## Test plan

- [x] `python -m pytest dandi/tests/test_brain_areas.py -v` — 29 unit tests pass
- [x] `python -m pytest dandi/tests/test_metadata.py -v -k anatomy` — 3 integration tests pass
- [x] `pre-commit run --files <all changed files>` — all hooks pass
- [x] CI passes

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by bendichter** [comment]
sister PR: https://github.com/NeurodataWithoutBorders/nwbinspector/pull/671

File: dandi/metadata/brain_areas.py [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+                for val in d.values():
+                    val = str(val).strip()
+                    if val and val.lower() not in _TRIVIAL_VALUES:
+                        # Skip purely numeric values (e.g. depth)
+                        try:
+                            float(val)
+                        except ValueError:
+                            tokens.append(val)
+                return tokens
+        except (ValueError, SyntaxError):
```
[the remark this answers is not in the document]
Comment (yarikoptic) [reply dandi/metadata/brain_areas.py]: @bendichter something should be at least logged here or best to give example on when is it happening -- is that a broken structure from allen (report upstream?) or expected and unhandled explicitly above ?  may be it would signal need to just add one more _TRIVIAL_VALUES ?

overall -- unclear

File: dandi/metadata/brain_areas.py [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+        if name not in name_exact:
+            name_exact[name] = s
+        name_low = name.lower()
+        if name_low not in name_lower:
+            name_lower[name_low] = s
+    return acronym_exact, acronym_lower, name_exact, name_lower
+
+
+def _parse_location_string(location: str) -> list[str]:
+    """Parse a raw NWB location string into area tokens.
```
Comment (yarikoptic) [inline dandi/metadata/brain_areas.py]:
```suggestion
    """Parse a raw NWB location string into area tokens ignoring numerics etc.
```

File: dandi/metadata/util.py [hunk 3]
```diff
@@ -552,22 +554,65 @@ def func(metadata: dict) -> list[M]:
...
+    from .brain_areas import locations_to_anatomy
+
+    locations = metadata.get("brain_locations")
+    if not locations:
+        return []
+
+    # Only apply Allen CCF matching for mouse (NCBITaxon_10090)
+    species = metadata.get("species")
+    if species is None:
+        return []
```
Comment (yarikoptic) [inline dandi/metadata/util.py]: ask your agent to be python 3.8+ and imho we can be generic:

```suggestion
    if not (species := metadata.get("species")):
        return []
```

File: dandi/metadata/util.py [hunk 4]
```diff
@@ -552,22 +554,65 @@ def func(metadata: dict) -> list[M]:
...
+    species_str = str(species).lower()
+    is_mouse = (
+        "10090" in species_str
+        or "mus musculus" in species_str
+        or "mouse" in species_str
+    )
```
Comment (yarikoptic) [inline dandi/metadata/util.py]: we have matching for species elsewhere -- reuse here if you can about matching to mice.

File: dandi/pynwb_utils.py [hunk 5]
```diff
@@ -384,6 +387,74 @@
...
+            # Might be a DynamicTable in some NWB versions
+            try:
+                col_names = getattr(ic_electrodes, "colnames", ())
+                if col_names and "location" in col_names:
+                    for val in ic_electrodes["location"].data[:]:
+                        if isinstance(val, bytes):
+                            val = val.decode("utf-8", errors="replace")
+                        if val and isinstance(val, str):
+                            locations.append(val)
+            except Exception:
```
[the remark this answers is not in the document]
Comment (yarikoptic) [reply dandi/pynwb_utils.py]: too wide -- not even log message... RF to become more specific, and add logging here even if at .debug level.  Otherwise we might endup wild chasing bugs later here

**Review by yarikoptic — CHANGES_REQUESTED** [review]
I think it would work but we need to make it a little "better", in particular a little more specific to how it processes user entered data found in NWB so we do not end up chasing silently "handled" bugs

**Review by yarikoptic — CHANGES_REQUESTED** [review]
## Review of PR #1807: Extract Brain Area Anatomy from NWB Location Fields

### Summary

This PR adds the ability to extract brain region locations from NWB files (imaging planes, electrodes table, intracellular electrodes), match them against a bundled Allen Mouse Brain Atlas CCF structure database (~94KB, 1,327 structures), and populate `BioSample.anatomy` within the `wasDerivedFrom` chain. The feature is mouse-only (NCBITaxon_10090).

The initial review raised valid concerns around broad exception handling, non-specific filtering, inflexible dict key matching, hardcoded species logic, verbose tests, and missing logging. The follow-up commit d6b5be1 addresses most of these. Below is an assessment of what was done well, what was partially addressed, and what still warrants attention.

---

### Review Comments: How Well Were They Addressed?

**1. Bare except / missing logging** — Addressed adequately in `brain_areas.py` (now has `lgr.debug(...)`) and the icephys block in `pynwb_utils.py`. However, two remaining `except Exception:` blocks in the ecephys section of `_get_brain_locations()` still swallow exceptions silently:

```python
try:
    col_names = electrodes.colnames
except Exception:
    col_names = ()
```
and:
```python
try:
    locs = electrodes["location"].data[:]
except Exception:
    locs = []
```

These should get `lgr.debug(...)` treatment consistent with the icephys block. The concern about "wild chasing bugs" applies equally here.

**2. Non-specific numeric filtering** — Addressed. Extracted into `_is_numeric()` helper. The `float()` approach is more robust than regex for this. Acceptable resolution.

**3. Dict key matching** — Well addressed. `_normalise_key()` strips `[\s_-]` and lowercases, plus expanded `_AREA_KEYS` includes `areaname`, `brainarea`, `brainregion`, `regionname`, `name`. Matches the spirit of the review comment.

**4. Reusing species matching logic** — Partially addressed. `_is_mouse()` now references `species_map[0]` instead of hardcoding strings, which is better. But it relies on a positional index, which is fragile — if someone reorders `species_map`, mouse detection breaks silently. Safer alternatives:
- Look up by a stable key (e.g., filter `species_map` for the entry containing `"10090"`)
- Or add an assertion that `species_map[0]` is the mouse entry

**5. Parametric tests** — Well addressed. Tests are clean with `@pytest.mark.parametrize`. The file went from ~142 lines to 107, maintaining coverage while being more idiomatic.

**6. Exception handling in pynwb_utils.py** — Partially addressed. As noted in point 1, only the icephys block got logging. The ecephys block retains two bare `except Exception:` without logging.

---

### Additional Observations

#### Architecture and Design

**Placement of anatomy in `wasDerivedFrom`** — Attaching anatomy to the *deepest* (first-created, i.e., tissue-level) `BioSample` is conceptually sound. The fallback creating a synthetic `BioSample(identifier="brain-region-sample")` when no tissue/slice/cell chain exists is reasonable but deserves a comment explaining why this identifier was chosen and whether it could collide or be meaningful downstream.

**The `deepest` variable tracking in `extract_wasDerivedFrom()`** — The loop builds the chain bottom-up (tissue -> slice -> cell), capturing `deepest` as the first sample created. Correct but subtle. A short comment clarifying the chain direction would help future readers.

**`_extract_brain_anatomy()` uses a lazy import** — `from .brain_areas import locations_to_anatomy` inside the function body. This avoids loading the 94KB JSON for non-mouse files. Reasonable.

#### Data Concerns

**Bundled `allen_ccf_structures.json`** — The regeneration script fetches from `http://api.brain-map.org` (HTTP, not HTTPS). Consider:
- A version/date stamp in the JSON or a comment for provenance
- Verify whether HTTPS is available and prefer it
- No documentation on *when* to regenerate (e.g., linked to Allen API release cycles)

#### Code Quality

**`_parse_location_string` has duplicated fallback logic** — Both the dict-literal branch and the key-value branch have nearly identical patterns:
```python
tokens = []
for v in d.values():  # or kv.values()
    v = str(v).strip()
    if v and v.lower() not in _TRIVIAL_VALUES and not _is_numeric(v):
        tokens.append(v)
```
Could be extracted to a small helper like `_non_trivial_values(d)`.

**The key-value regex `r"(\w+)\s*:\s*(.+)"` is greedy** — The `"://"` guard protects against URLs as the *entire* location string, but not against URLs inside values of a key-value string. Edge case, but worth noting.

**`"brain"` in `_TRIVIAL_VALUES`** — Intentional (uninformative), but should `"cortex"` or other overly-broad terms be included? Worth considering for future refinement.

#### Test Coverage

Unit tests in `test_brain_areas.py` are solid — simple strings, dict literals, key-value, comma-separated, trivial values, case-insensitive matching, deduplication.

Integration tests cover: mouse with electrodes, rat skipped, existing BioSample chain.

**Missing test cases worth considering:**
- Imaging plane locations (ophys path) — currently only ecephys is integration-tested
- The synthetic `"brain-region-sample"` BioSample fallback (mouse with locations but no tissue/slice/cell ids)
- `_is_mouse()` edge cases (full URI, just `"Mus musculus"`, just `"mouse"`)
- `brain_locations` containing *only* unmatched tokens for a mouse

#### Minor

- `_build_lookup_dicts` return type annotation wraps oddly across lines
- `_structure_to_anatomy` lacks a docstring (though self-explanatory)
- Both `pyproject.toml` codespell skip and `.pre-commit-config.yaml` codespell exclude add the same JSON file — redundant but harmless

---

### Items to Address Before Merge

1. **Add `lgr.debug(...)` to the two remaining silent `except Exception:` blocks in `_get_brain_locations()` ecephys section** — this was the core initial review feedback and it's inconsistent that only icephys got it
2. **Make `_is_mouse()` less fragile** — at minimum add an assertion or comment that `species_map[0]` is mouse, or look it up by a stable criterion
3. **Consider HTTPS for the Allen API URL** in `generate_allen_structures.py`

**Review by bendichter — COMMENTED** [review]
Thanks for the thorough review! All items have been addressed across three commits:

### Items to Address Before Merge

**1. Add `lgr.debug(...)` to ecephys except blocks** — Done. Both `except Exception:` blocks in `_get_brain_locations()` (for `electrodes.colnames` and `electrodes["location"].data[:]`) now log at debug level with `exc_info=True`, consistent with the icephys block.

**2. Make `_is_mouse()` less fragile** — Done. Instead of relying on `species_map[0]`, now looks up the mouse entry by filtering for `"10090"` in the URI: `next((entry for entry in species_map if "10090" in entry[2]), None)`.

**3. HTTPS for Allen API URL** — Done. Changed to `https://api.brain-map.org/...` in `generate_allen_structures.py`.

### Additional Observations Addressed

- **Duplicated fallback logic** — Extracted into `_non_trivial_values(d)` helper, used by both the dict-literal and key-value branches in `_parse_location_string`.
- **Comment on `brain-region-sample`** — Added comment explaining it's a fixed placeholder that won't collide with UUID-based NWB object IDs.
- **Comment on chain direction** — Added comment on `deepest` explaining the tissue → slice → cell iteration order and why anatomy attaches to the first (tissue-level) sample.
- **`_structure_to_anatomy` docstring** — Added.
- **`_build_lookup_dicts` return type formatting** — Fixed to use the `] ->` style instead of the odd `-> (tuple[...])` wrapping.

### Missing Test Cases Added

- **Ophys integration test** — Creates an NWB file with an imaging plane and verifies anatomy extraction via the ophys path.
- **Synthetic `brain-region-sample` fallback** — Mouse with brain locations but no tissue/slice/cell IDs produces a synthetic BioSample with correct anatomy.
- **`_is_mouse()` edge cases** — Parametrized positive cases (full URI, `"Mus musculus"`, `"mouse"`, lowercase) and negative cases (rat, human, Drosophila, empty string, missing species key).
- **Unmatched tokens only** — Mouse with only non-matching location strings produces no anatomy and no synthetic BioSample.

File: dandi/tests/test_metadata.py [hunk 6]
```diff
@@ -707,6 +711,248 @@ def test_session_duration_with_events(tmp_path: Path) -> None:
...
+        excitation_lambda=600.0,
+        indicator="GCaMP6f",
+        location="Primary visual area",
+    )
+
+    with NWBHDF5IO(str(nwb_path), "w") as io:
+        io.write(nwbfile)
+
+    from ..metadata.nwb import nwb2asset
+    from ..misctypes import DUMMY_DANDI_ETAG
```
[the remark this answers is not in the document]
Comment (bendichter) [reply dandi/tests/test_metadata.py]: Fixed — removed all 9 redundant in-function imports (they were already at the top of the file). Also added "Never place imports inside functions or methods; always use top-level imports" to CLAUDE.md so this won't happen again. (08ff314c)

File: pyproject.toml [hunk 7]
```diff
@@ -161,7 +161,7 @@ tag_prefix = ""
 parentdir_prefix = ""
 
 [tool.codespell]
-skip = "_version.py,due.py,versioneer.py,*.vcr.yaml,venv,venvs,pyproject.toml"
+skip = "_version.py,due.py,versioneer.py,*.vcr.yaml,venv,venvs,pyproject.toml,allen_ccf_structures.json"
```
[the remark this answers is not in the document]
Comment (bendichter) [reply pyproject.toml:178]: I think skipping the file is appropriate here. "accesory" is a typo but there isn't a clear repo where I can raise it. Regardless, it's out of scope for this PR
